## Week 4, Lab 1: Predicting Left-Handedness from Psychological Factors
> Author: Matt Brems

We can sketch out the data science process as follows:
1. Define the problem.
2. Obtain the data.
3. Explore the data.
4. Model the data.
5. Evaluate the model.
6. Answer the problem.

We'll walk through a full data science problem in this lab. 

---
## Step 1: Define The Problem.

You're currently a data scientist working at a university. A professor of psychology is attempting to study the relationship between personalities and left-handedness. They have tasked you with gathering evidence so that they may publish.

Specifically, the professor says "I need to prove that left-handedness is caused by some personality trait. Go find that personality trait and the data to back it up."

As a data scientist, you know that any real data science problem must be **specific** and **conclusively answerable**. For example:
- Bad data science problem: "What is the link between obesity and blood pressure?"
    - This is vague and is not conclusively answerable. That is, two people might look at the conclusion and one may say "Sure, the problem has been answered!" and the other may say "The problem has not yet been answered."
- Good data science problem: "Does an association exist between obesity and blood pressure?"
    - This is more specific and is conclusively answerable. The problem specifically is asking for a "Yes" or "No" answer. Based on that, two independent people should both be able to say either "Yes, the problem has been answered" or "No, the problem has not yet been answered."
- Excellent data science problem: "As obesity increases, how does blood pressure change?"
    - This is very specific and is conclusively answerable. The problem specifically seeks to understand the effect of one variable on the other.

### 1. In the context of the left-handedness and personality example, what are three specific and conclusively answerable problems that you could answer using data science? 

> You might find it helpful to check out the codebook in the repo for some inspiration.

Answer:
1. As of people studied how to win gambling increases, how does the left-handedness change?
2. Which personal trait have highest positive correlation with left-handedness?
3. Does left-handedness affected by how people like guns?

---
## Step 2: Obtain the data.

### 2. Read in the file titled "data.csv."
> Hint: Despite being saved as a .csv file, you won't be able to simply `pd.read_csv()` this data!

In [4]:
import pandas as pd
df = pd.read_table('data.csv')
df.head()

Q1  Q2  Q3  Q4  Q5  Q6  Q7  Q8  Q9  Q10  ...  country  fromgoogle  engnat  \
0   4   1   5   1   5   1   5   1   4    1  ...       US           2       1   
1   1   5   1   4   2   5   5   4   1    5  ...       CA           2       1   
2   1   2   1   1   5   4   3   2   1    4  ...       NL           2       2   
3   1   4   1   5   1   4   5   4   3    5  ...       US           2       1   
4   5   1   5   1   5   1   5   1   3    1  ...       US           2       1   

   age  education  gender  orientation  race  religion  hand  
0   22          3       1            1     3         2     3  
1   14          1       2            2     6         1     1  
2   30          4       1            1     1         1     2  
3   18          2       2            5     3         2     2  
4   22          3       1            1     3         2     3  

[5 rows x 56 columns]

### 3. Suppose that, instead of us giving you this data in a file, you were actually conducting a survey to gather this data yourself. From an ethics/privacy point of view, what are three things you might consider when attempting to gather this data?
> When working with sensitive data like sexual orientation or gender identity, we need to consider how this data could be used if it fell into the wrong hands!

Answer:
1. Collect data anonymously
2. Considering a 'prefer not to answer' option for some sensitive questions.
3. Not collecting name for survey incase of the data leakage.

---
## Step 3: Explore the data.

### 4. Conduct exploratory data analysis on this dataset.
> If you haven't already, be sure to check out the codebook in the repo, as that will help in your EDA process.

In [5]:
df.columns

Index(['Q1', 'Q2', 'Q3', 'Q4', 'Q5', 'Q6', 'Q7', 'Q8', 'Q9', 'Q10', 'Q11',
       'Q12', 'Q13', 'Q14', 'Q15', 'Q16', 'Q17', 'Q18', 'Q19', 'Q20', 'Q21',
       'Q22', 'Q23', 'Q24', 'Q25', 'Q26', 'Q27', 'Q28', 'Q29', 'Q30', 'Q31',
       'Q32', 'Q33', 'Q34', 'Q35', 'Q36', 'Q37', 'Q38', 'Q39', 'Q40', 'Q41',
       'Q42', 'Q43', 'Q44', 'introelapse', 'testelapse', 'country',
       'fromgoogle', 'engnat', 'age', 'education', 'gender', 'orientation',
       'race', 'religion', 'hand'],
      dtype='object')

In [6]:
df.describe(include = 'all')

Q1           Q2           Q3           Q4           Q5  \
count   4184.000000  4184.000000  4184.000000  4184.000000  4184.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean       1.962715     3.829589     2.846558     3.186902     2.865440   
std        1.360291     1.551683     1.664804     1.476879     1.545798   
min        0.000000     0.000000     0.000000     0.000000     0.000000   
25%        1.000000     3.000000     1.000000     2.000000     1.000000   
50%        1.000000     5.000000     3.000000     3.000000     3.000000   
75%        3.000000     5.000000     5.000000     5.000000     4.000000   
max        5.000000     5.000000     5.000000     5.000000     5.000000   

                 Q6           Q7           Q8           Q9          Q10  ...  \
count   4184.000000  4184.000000  4184.000000  4184.000000  4184.000000  ...   
unique          NaN          NaN          NaN          NaN          NaN  ...   
top             NaN          NaN          NaN          NaN          NaN  ...   
freq            NaN          NaN          NaN          NaN          NaN  ...   
mean       3.672084     3.216539     3.184512     2.761233     3.522945  ...   
std        1.342238     1.490733     1.387382     1.511805     1.242890  ...   
min        0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%        3.000000     2.000000     2.000000     1.000000     3.000000  ...   
50%        4.000000     3.000000     3.000000     3.000000     4.000000  ...   
75%        5.000000     5.000000     4.000000     4.000000     5.000000  ...   
max        5.000000     5.000000     5.000000     5.000000     5.000000  ...   

        country   fromgoogle       engnat           age    education  \
count      4184  4184.000000  4184.000000   4184.000000  4184.000000   
unique       94          NaN          NaN           NaN          NaN   
top          US          NaN          NaN           NaN          NaN   
freq       2468          NaN          NaN           NaN          NaN   
mean        NaN     1.576243     1.239962     30.370698     2.317878   
std         NaN     0.494212     0.440882    367.201726     0.874264   
min         NaN     1.000000     0.000000     13.000000     0.000000   
25%         NaN     1.000000     1.000000     18.000000     2.000000   
50%         NaN     2.000000     1.000000     21.000000     2.000000   
75%         NaN     2.000000     1.000000     27.000000     3.000000   
max         NaN     2.000000     2.000000  23763.000000     4.000000   

             gender  orientation         race     religion         hand  
count   4184.000000  4184.000000  4184.000000  4184.000000  4184.000000  
unique          NaN          NaN          NaN          NaN          NaN  
top             NaN          NaN          NaN          NaN          NaN  
freq            NaN          NaN          NaN          NaN          NaN  
mean       1.654398     1.833413     5.013623     2.394359     1.190966  
std        0.640915     1.303454     1.970996     2.184164     0.495357  
min        0.000000     0.000000     0.000000     0.000000     0.000000  
25%        1.000000     1.000000     5.000000     1.000000     1.000000  
50%        2.000000     1.000000     6.000000     2.000000     1.000000  
75%        2.000000     2.000000     6.000000     2.000000     1.000000  
max        3.000000     5.000000     7.000000     7.000000     3.000000  

[11 rows x 56 columns]

In [7]:
df.isnull().sum()

Q1             0
Q2             0
Q3             0
Q4             0
Q5             0
Q6             0
Q7             0
Q8             0
Q9             0
Q10            0
Q11            0
Q12            0
Q13            0
Q14            0
Q15            0
Q16            0
Q17            0
Q18            0
Q19            0
Q20            0
Q21            0
Q22            0
Q23            0
Q24            0
Q25            0
Q26            0
Q27            0
Q28            0
Q29            0
Q30            0
Q31            0
Q32            0
Q33            0
Q34            0
Q35            0
Q36            0
Q37            0
Q38            0
Q39            0
Q40            0
Q41            0
Q42            0
Q43            0
Q44            0
introelapse    0
testelapse     0
country        0
fromgoogle     0
engnat         0
age            0
education      0
gender         0
orientation    0
race           0
religion       0
hand           0
dtype: int64

---
## Step 4: Model the data.

### 5. Suppose I wanted to use Q1 - Q44 to predict whether or not the person is left-handed. Would this be a classification or regression problem? Why?

Answer: Classification, we are predicting whether or not the person is left-handed.

### 6. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed based on their responses to Q1 - Q44. Before doing that, however, you remember that it is often a good idea to standardize your variables. In general, why would we standardize our variables? Give an example of when we would standardize our variables.

Answer: Standardize variables help prevent prediction become bias whereby different predictor without stadardization have different weightage. Example without stadarize the kilograms and grams, if we increase 1 for both. 1kg become 2kg and 1000g become 1001g there's a big difference.

### 7. Give an example of when we might not standardize our variables.

Answer: If our variables are same scale, we dont have to standardize them.

### 8. Based on your answers to 6 and 7, do you think we should standardize our predictor variables in this case? Why or why not?

Answer: No need, they are using same scale (1-5)

### 9. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed. What munging/cleaning do we need to do to our $y$ variable in order to explicitly answer this question? Do it.

Answer: 
1. Check all the data if there any other 'weird data'
2. Remove the unwanted data (0)

In [8]:
df['hand'].value_counts()

hand
1    3542
2     452
3     179
0      11
Name: count, dtype: int64

In [9]:
#dropping missing value from hand column
df = df[df['hand'] != 0].reset_index()

In [10]:
df['hand'].value_counts()

hand
1    3542
2     452
3     179
Name: count, dtype: int64

In [13]:
# adding new column y to only show 1 when left-handed
df['y'] = [1 if i == 2 else 0 for i in df['hand']]
df['y'].value_counts()

y
0    3721
1     452
Name: count, dtype: int64

### 10. The professor for whom you work suggests that you set $k = 4$. In this specific case, why might this be a bad idea?

Answer: Ideally, k number is expected to be an odd number to prevent the conflict if there might be a tie when predicting the discrete output

### 11. Let's *(finally)* use $k$-nearest neighbors to predict whether or not a person is left-handed!

> Be sure to create a train/test split with your data!

> Create four separate models, one with $k = 3$, one with $k = 5$, one with $k = 15$, and one with $k = 25$.

> Instantiate and fit your models.

In [14]:
X = df.drop(columns=['index', 'introelapse', 'testelapse', 'country',
       'fromgoogle', 'engnat', 'age', 'education', 'gender', 'orientation',
       'race', 'religion', 'hand', 'y'], axis = 1)

y = df['y']

In [17]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [18]:
k_3 = KNeighborsClassifier(n_neighbors = 3)
k_3.fit(X_train, y_train)

k_5 = KNeighborsClassifier(n_neighbors = 5)
k_5.fit(X_train, y_train)

k_15 = KNeighborsClassifier(n_neighbors = 15)
k_15.fit(X_train, y_train)

k_25 = KNeighborsClassifier(n_neighbors = 25)
k_25.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=25)

Being good data scientists, we know that we might not run just one type of model. We might run many different models and see which is best.

### 12. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, let's check the [documentation for logistic regression in sklearn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html). Is there default regularization? If so, what is it? If not, how do you know?

Answer: 
1. Yes there is default regularization.
For example, default penalty is L2 (Ridge)
C = 1.0 indicates that the inverse of our regularization strength is 1.
2. There is also non default which is the random_state based on the documentation.

### 13. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, should we standardize our features?

Answer: Do not have, we have drop other features that have different scale value.

### 14. Let's use logistic regression to predict whether or not the person is left-handed.


> Be sure to use the same train/test split with your data as with your $k$-NN model above!

> Create four separate models, one with LASSO and $\alpha = 1$, one with LASSO and $\alpha = 10$, one with Ridge and $\alpha = 1$, and one with Ridge and $\alpha = 10$. *(Hint: Be careful with how you specify $\alpha$ in your model!)*

> Instantiate and fit your models.

In [19]:
from sklearn.linear_model import LogisticRegression

In [24]:
lasso_1 = LogisticRegression(penalty = 'l1', C = 1.0, solver = 'liblinear')
lasso_1.fit(X_train, y_train)

lasso_10 = LogisticRegression(penalty = 'l1', C = 0.1, solver = 'liblinear')
lasso_10.fit(X_train, y_train)

ridge_1 = LogisticRegression(penalty = 'l2', C = 1.0, solver = 'liblinear')
ridge_1.fit(X_train, y_train)

ridge_10 = LogisticRegression(penalty = 'l2', C = 0.1, solver = 'liblinear')
ridge_10.fit(X_train, y_train)

LogisticRegression(C=0.1, solver='liblinear')

---
## Step 5: Evaluate the model(s).

### 15. Before calculating any score on your data, take a step back. Think about your $X$ variable and your $Y$ variable. Do you think your $X$ variables will do a good job of predicting your $Y$ variable? Why or why not? What impact do you think this will have on your scores?

Answer: Only by visualizing data, we can see if there is trend, and by increasing number of survey, we can prove that our statement is more valid but not certainly 100%

### 16. Using accuracy as your metric, evaluate all eight of your models on both the training and testing sets. Put your scores below. (If you want to be fancy and generate a table in Markdown, there's a [Markdown table generator site linked here](https://www.tablesgenerator.com/markdown_tables#).)
- Note: Your answers here might look a little weird. You didn't do anything wrong; that's to be expected!

In [25]:
print("k-nearest neighbors training accuracy with k = 3: " + str(k_3.score(X_train, y_train)))
print("k-nearest neighbors testing accuracy with k = 3: " + str(k_3.score(X_test, y_test)))

print("k-nearest neighbors training accuracy with k = 5: " + str(k_5.score(X_train, y_train)))
print("k-nearest neighbors testing accuracy with k = 5: " + str(k_5.score(X_test, y_test)))

print("k-nearest neighbors training accuracy with k = 15: " + str(k_15.score(X_train, y_train)))
print("k-nearest neighbors testing accuracy with k = 15: " + str(k_15.score(X_test, y_test)))

print("k-nearest neighbors training accuracy with k = 25: " + str(k_25.score(X_train, y_train)))
print("k-nearest neighbors testing accuracy with k = 25: " + str(k_25.score(X_test, y_test)))

print("logistic regression training accuracy with LASSO penalty, alpha = 1: " + str(lasso_1.score(X_train, y_train)))
print("logistic regression testing accuracy with LASSO penalty, alpha = 1: " + str(lasso_1.score(X_test, y_test)))

print("logistic regression training accuracy with LASSO penalty, alpha = 10: " + str(lasso_10.score(X_train, y_train)))
print("logistic regression testing accuracy with LASSO penalty, alpha = 10: " + str(lasso_10.score(X_test, y_test)))

print("logistic regression training accuracy with Ridge penalty, alpha = 1: " + str(ridge_1.score(X_train, y_train)))
print("logistic regression testing accuracy with Ridge penalty, alpha = 1: " + str(ridge_1.score(X_test, y_test)))

print("logistic regression training accuracy with Ridge penalty, alpha = 10: " + str(ridge_10.score(X_train, y_train)))
print("logistic regression testing accuracy with Ridge penalty, alpha = 10: " + str(ridge_10.score(X_test, y_test)))

k-nearest neighbors training accuracy with k = 3: 0.9057206775327581
k-nearest neighbors testing accuracy with k = 3: 0.8582375478927203
k-nearest neighbors training accuracy with k = 5: 0.8900607222754874
k-nearest neighbors testing accuracy with k = 5: 0.882183908045977
k-nearest neighbors training accuracy with k = 15: 0.8906999041227229
k-nearest neighbors testing accuracy with k = 15: 0.8936781609195402
k-nearest neighbors training accuracy with k = 25: 0.8910194950463407
k-nearest neighbors testing accuracy with k = 25: 0.8936781609195402
logistic regression training accuracy with LASSO penalty, alpha = 1: 0.8913390859699585
logistic regression testing accuracy with LASSO penalty, alpha = 1: 0.8936781609195402
logistic regression training accuracy with LASSO penalty, alpha = 10: 0.8913390859699585
logistic regression testing accuracy with LASSO penalty, alpha = 10: 0.8936781609195402
logistic regression training accuracy with Ridge penalty, alpha = 1: 0.8913390859699585
logistic 

**Answer:**

|        Model        |    Value of $k$   | Penalty |   Value of $\alpha$   | Training Accuracy | Testing Accuracy |
|:-------------------:|:--------------:|:-------:|:----------:|:-----------------:|:----------------:|
|         $k$-nn         |  $k = 3$ |    NA   |     NA     |       0.9057      |      0.8573      |
|         $k$-nn         |  $k = 5$ |    NA   |     NA     |       0.8910      |      0.8812      |
|         $k$-nn         | $k = 15$ |    NA   |     NA     |       0.8907      |      0.8937      |
|         $k$-nn         | $k = 25$ |    NA   |     NA     |       0.8910      |      0.8937      |
| logistic regression |   NA   |  LASSO  |  $\alpha = 1$ |       0.8913      |      0.8937      |
| logistic regression |   NA   |  LASSO  | $\alpha = 10$ |       0.8913      |      0.8937      |
| logistic regression |   NA   |  Ridge  |  $\alpha = 1$ |       0.8913      |      0.8937      |
| logistic regression |   NA   |  Ridge  | $\alpha = 10$ |       0.8913      |      0.8937      |

### 17. In which of your $k$-NN models is there evidence of overfitting? How do you know?

Answer: k=3 and k=5 are overfitting, because the training accurancy > testing accurancy

### 18. Broadly speaking, how does the value of $k$ in $k$-NN affect the bias-variance tradeoff? (i.e. As $k$ increases, how are bias and variance affected?)

Answer:
- As k increases, our bias increases and our variance decreases.
- As k decreases, our bias decreases and our variance increases.

### 19. If you have a $k$-NN model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Answer:
1. Increase number of k to reduce the variance. 
2. Consider change to logistic regression
3. Reduce the number of questions, lower the chance of diferent question which same similar correlation.

### 20. In which of your logistic regression models is there evidence of overfitting? How do you know?

Answer:
None, all testing score are higher than training score

### 21. Broadly speaking, how does the value of $C$ in logistic regression affect the bias-variance tradeoff? (i.e. As $C$ increases, how are bias and variance affected?)

Answer: No affecting the bias-variance tradeoff in this case

But in general,
as C increases, we regularize less. If we regularize less, then our variance increases and our bias decreases; As C decreases, we regularize more. If we regularize more, then our variance decreases and our bias increases.

### 22. For your logistic regression models, play around with the regularization hyperparameter, $C$. As you vary $C$, what happens to the fit and coefficients in the model? What do you think this means in the context of this specific problem?

Answer: No changes even change the C = 10.0

### 23. If you have a logistic regression model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Answer:
1. remove not relavance features.
2. add more relavance features.
3. adjust the alpha value.

---
## Step 6: Answer the problem.

### 24. Suppose you want to understand which psychological features are most important in determining left-handedness. Would you rather use $k$-NN or logistic regression? Why?

Answer:
If the dataset is not excessively large, and interpretability and understanding the relative importance of psychological features are crucial, logistic regression would be preferred. However, if the relationship between features and left-handedness is highly nonlinear, and interpretability is less critical, k-NN might be a better choice for its flexibility and lack of assumptions about the data's underlying distribution.

### 25. Select your logistic regression model that utilized LASSO regularization with $\alpha = 1$. Interpret the coefficient for `Q1`.

In [26]:
lasso_1.coef_

array([[-0.01104202, -0.00583879, -0.01422173, -0.05968224,  0.03933292,
        -0.00041527, -0.00444172, -0.18039489, -0.0096738 ,  0.03044125,
        -0.00208929,  0.00467544, -0.06746771,  0.05779557, -0.07135083,
         0.04056304,  0.02741013, -0.02441668, -0.00115956, -0.06537179,
        -0.06411849, -0.09501479, -0.01627524, -0.03430875,  0.04931216,
         0.09823913,  0.03976372, -0.0439046 ,  0.03584252,  0.01001705,
         0.0057564 ,  0.01019485, -0.0416222 , -0.01805979,  0.03861331,
        -0.02891216, -0.032813  ,  0.07534692, -0.06346307, -0.11185345,
        -0.0328195 , -0.05797544, -0.1147471 ,  0.03272759]])

The coefficient for `Q1` is -0.01104.

In [27]:
import numpy as np

np.exp(lasso_1.coef_[0][0])

0.9890187166088636

We can interpret the coefficient for `Q1` in one of three ways.
1. As the value for `Q1` increases by 1, the log-odds of being left-handed decreases by 0.01104.
2. As the value for `Q1` increases by 1, an individual is 98.90% as likely to be left-handed.
3. As the value for `Q1` increases by 1, an individual is 1.1% less likely to be left-handed.

### 26. If you have to select one model overall to be your *best* model, which model would you select? Why?
- Usually in the "real world," you'll fit many types of models but ultimately need to pick only one! (For example, a client may not understand what it means to have multiple models, or if you're using an algorithm to make a decision, it's probably pretty challenging to use two or more algorithms simultaneously.) It's not always an easy choice, but you'll have to make it soon enough. Pick a model and defend why you picked this model!

Answer: Logistic regression model

### 27. Circle back to the three specific and conclusively answerable questions you came up with in Q1. Answer one of these for the professor based on the model you selected!

Answer:
1. As of people studied how to win gambling increases, how does the left-handedness change?
2. Which personal trait have highest positive correlation with left-handedness?
3. Does left-handedness affected by how people like guns?


- I would pick logistic regression model


### BONUS:
Looking for more to do? Probably not - you're busy! But if you want to, consider exploring the following. (They could make for a blog post!)
- Create a visual plot comparing training and test metrics for various values of $k$ and various regularization schemes in logistic regression.
- Rather than just evaluating models based on accuracy, consider using sensitivity, specificity, etc.
- In the context of predicting left-handedness, why are unbalanced classes concerning? If you were to re-do this process given those concerns, what changes might you make?
- Fit and evaluate a generalized linear model other than logistic regression (e.g. Poisson regression).
- Suppose this data were in a `SQL` database named `data` and a table named `inventory`. What `SQL` query would return the count of people who were right-handed, left-handed, both, or missing with their class labels of 1, 2, 3, and 0, respectively? (You can assume you've already logged into the database.)